### ISSUE TITLE
Stuttering playback when restarting music playback after laptop is closed

### ISSUE THREAD

**Issue by dasl-** [mikebrady/shairport-sync#1496]
Hi, I'm getting stuttering playback in certain conditions:

1) Play music via macOS Music app (or youtube, and possibly other music playback apps) through shairport sync
2) close macbook lid for ~20 seconds
3) open macbook, resume music playback
4) observe constant stuttering of music playback

Disconnecting and reconnecting to airplay resolves the stuttering issues.

These are the logs when (4) occurs: https://gist.github.com/dasl-/cbdbe8537304be02ffe6e49ea65a2d5e

Software versions:
```
% nqptp -V ; shairport-sync -V
Version: 1.1-dev-156-gc13a5f0-dirty. Shared Memory Interface Version: 7.
4.1-dev-406-g82956dc7-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc

# raspberry pi 4
% uname -a
Linux piwall10 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l GNU/Linux
```

Shairport sync config: https://gist.github.com/dasl-/496db91d598872e000db28744fad2573

I am not 100% sure, but I think this may have started happening only after I recently upgraded NQPTP and shairport sync. I upgraded both on July 4, because I saw activity on this github issue: https://github.com/mikebrady/shairport-sync/issues/1484

I occasionally experienced problems similar to the author of that issue. I don't have enough data yet to confirm that your changes have solved that issue, but thus far I have not noticed the issue described in #1484  since upgrading (time and additional usage will tell if it recurs...): "if I wait 4-5 minutes and try to play music again, the timeline starts counting down the seconds but there is no sound"

So all that to say: it's possible that your recent fixes solved one problem and caused another problem? But again, I'm not certain about this. Thanks for all your work on SPS! It's inspiring :)

**Comment by mikebrady** [comment]
Thanks for the report. I’ll take a look!

**Comment by mikebrady** [comment]
Also, is it an Apple Silicon MacBook or an Intel MacBook?

**Comment by dasl-** [comment]
>Just looking at the log file, I see that it's a realtime stream, which would eliminate macOS Music as a source. Could it have been YouTube or Spotify, possibly?

Good point. I did some investigating. I typically connect to airplay via the Sound item in the system menu bar:
![image](https://user-images.githubusercontent.com/1332576/178017141-a305db96-d92d-4df2-bd46-d9c36586b3c9.png)

I've observed that when I do this, it uses a realtime stream. So I was playing music via the Music app, but connecting in a manner that resulted in using a realtime stream.

When I connect via the Music app airplay icon, it uses a buffered stream. I only observe this bug when connecting via the realtime stream method.

>Also, is it an Apple Silicon MacBook or an Intel MacBook?

Apple silicon M1

**Comment by dasl-** [comment]
Interestingly, I find that if I connect to shairport sync via the Music app airplay icon, if I try to play sound from another app, i.e. youtube in browser, it will sometimes not work. I would expect it to play youtube from the laptop speakers and Music from airplay. Furthermore, sometimes the Music app freezes (spinning beachball). Let me know if you'd like more details and I can perhaps file a separate bug report.

**Comment by mikebrady** [comment]
Thanks. I have a feeling that this is an Apple problem, but if you can get it to happen reliably, I’d be interested alright.

**Comment by mikebrady** [comment]
Can't get this to happen on an Intel Mac right now. It'll be a couple of days before I can try it on an M1 MacBook Pro, and I think the type of Mac might be important.

**Comment by dasl-** [comment]
RE my original bug report of stuttering playback, I can confirm that this only seems to happen on my M1 macbook. I'm able to consistently reproduce there, but on older generation I'm unable to reproduce. So I agree, it might be an apple problem. 

RE my [previous comment](https://github.com/mikebrady/shairport-sync/issues/1496#issuecomment-1179095691) about a separate issue, I'm unable to consistently reproduce, but if I figure out how I'll file a separate bug report.

**Comment by mikebrady** [comment]
Hello again. I have been able to get this to happen once or twice. There's definitely something there alright -- thanks for the report.

**Comment by mikebrady** [comment]
Update -- FWIW, a version of this problem also occurs with Apple's own AirPort Express. Nevertheless, it's not clear whether it's an Apple bug or an expected behaviour going forward (and leaving compatibility with the poor old AirPort Express behind). Time to do a bit more investigation...

**Comment by dasl-** [comment]
Out of curiosity, are you only able to reproduce on an M1 mac?

**Comment by mikebrady** [comment]
> Out of curiosity, are you only able to reproduce on an M1 Mac?

Yes, not on an Intel Mac, only on an M1 Mac -- I don't have access to an M2 Mac, but I guess it'll be the same.

Of course it's all guesswork, but my guess is that the entire audio system has been rearchitected and rebuilt for Apple Silicon, and that quite a lot of it is (or will be) shared across platforms.

**Comment by mikebrady** [comment]
There's an update to NQPTP in the `development` branch with three changes in it and it fixes this problem for me. Two of the changes are definite enhancements (so thanks for the report!) and one is a slightly hacky way of reviving the Mac's PTP clock after sleep. 

If you got a chance to try it out and report back, it would be appreciated. (Please remember to switch to the `development` branch. And also, remember to restart `shairport-sync` after you have rebuilt and reinstalled `nqptp`. Thanks.)

I still can't figure out if the M1's behaviour here is buggy or just expected new behaviour...

**Comment by dasl-** [comment]
Amazing! I won't be able to try this until I'm back home next week, but
I'll let you know then.

Regarding whether this is expected behavior- I must be missing something
because surely you don't mean that stuttering when awakening from sleep
should be expected behavior?

**Comment by dasl-** [comment]
I've just been running some tests on the new code. 

Versions I'm running:
```
% nqptp -V ; shairport-sync -V
Version: 1.1-dev-165-g794955b-dirty. Shared Memory Interface Version: 7.
4.1-dev-428-g9c8a3328-AirPlay2-OpenSSL-Avahi-ALSA-soxr-metadata-dbus-sysconfdir:/etc
```

I've tested this scenario maybe ~10 times, and I was only able to reproduce the issue once. So it's definitely an improvement (this issue is a minor inconvenience at most). Most times I tested, the laptop lid was closed for about 30 seconds. The one time I was able to reproduce the issue, the laptop lid was probably closed for about 5 minutes. 

After getting the issue to occur once, I tried to make it occur again by closing the laptop lid for exactly 5 minutes and exactly 10 minutes. Neither of these single attempts managed to trigger the issue again.

Logs from the one time I was able to reproduce: https://gist.github.com/dasl-/7cb286976c0f3cf39f56b0f9ce93e30f

**Comment by mikebrady** [comment]
Many thanks for your testing!

**Comment by mikebrady** [comment]
So, I’m thinking I might leave this until the new operating systems are out — I suspect there’ll be changes. Would that be okay?

**Comment by mikebrady** [comment]
Still playing around with this and making a little headway. BTW, I have noticed that if you use a HomePod mini as the AirPlay output device in place of the Shairport Sync device, the output connection is broken when the laptop is opened again. Sometimes, play reverts to the laptop speakers, but more often the audio seems to simply disappear. So it's clearly still a work-in-progress and not just a Shairport Sync thing.

**Comment by dasl-** [comment]
interesting, thanks
